src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Regex.cs

```diff
@@ -387,7 +486,87 @@ internal void Run<TState>(string input, int startat, ref TState state, MatchCall
             RegexRunner runner = Interlocked.Exchange(ref _runner, null) ?? CreateRunner();
             try
             {
-                runner.ScanInternal(this, input, startat, ref state, callback, reuseMatchObject, internalMatchTimeout);
+                runner.InitializeTimeout(internalMatchTimeout);
+                int runtextpos = startat;
+                while (true)
+                {
+                    runner.InitializeForScan(this, input, startat, false);
+                    runner.runtext = input;
```

Comment:
Can this be moved out of the while loop?

Reply:
It could, but the reason why it is inside now is because InternalPerformScan is nulling out runtext after its done (since this logic is shared with the other Run() method) so now we need to re-set it every iteration of the loop. We could alternatively add a bool parameter into InternalPerformScan which conditions if we should null it out or not depending on who is calling it if that is preferred.